File: python/score-after-flipping-matrix.py
class Solution:

    def get_row_score(row: List[int]) -> int:
        row_score = 0
        for i in range(len(row)):
            row_score = (row_score << 1) | row[i]
        return row_score


    def get_score(grid: List[List[int]]) -> int:
        score = 0
        for row in grid:
            score += Solution.get_row_score(row)
        return score


    def toggle_row(row: List[int]):
        for i in range(len(row)):
            row[i] ^= 1


    def maximize_rows(grid: List[List[int]]):
        for row in grid:
            if row[0] == 0:
                Solution.toggle_row(row)


    def toggle_column(grid: List[List[int]], i: int):
        for j in range(len(grid)):
            grid[j][i] ^= 1


    def maximize_columns(grid: List[List[int]]):
        row_num = len(grid)
        for i in range(len(grid[0])):
            one_count = sum([grid[j][i] for j in range(row_num)])
            zero_count = row_num - one_count
            if one_count < zero_count:
                Solution.toggle_column(grid, i)


    def matrixScore(self, grid: List[List[int]]) -> int:
        Solution.maximize_rows(grid)
        Solution.maximize_columns(grid)
        return Solution.get_score(grid)

File: go/score-after-flipping-matrix.go
func FlipRow(grid [][]int, i int) {
	for j := 0; j < len(grid[i]); j++ {
		grid[i][j] ^= 1
	}
}

func MaximizeRows(grid [][]int) {
	for i := 0; i < len(grid); i++ {
		if grid[i][0] == 0 {
			FlipRow(grid, i)
		}
	}
}

func SumCol(grid [][]int, j int) int {
	oneCount := 0
	for i := 0; i < len(grid); i++ {
		oneCount += grid[i][j]
	}
	return oneCount
}

func FlipCol(grid [][]int, j int) {
	for i := 0; i < len(grid); i++ {
		grid[i][j] ^= 1
	}
}

func MaximizeCols(grid [][]int) {
	for j := 0; j < len(grid[0]); j++ {
		oneCount := SumCol(grid, j)
		zeroCount := len(grid) - oneCount
		if oneCount < zeroCount {
			FlipCol(grid, j)
		}
	}
}

func GetRowScore(grid [][]int, i int) int {
	rowScore := 0
	for j := 0; j < len(grid[i]); j++ {
		rowScore = (rowScore << 1) | grid[i][j]
	}
	return rowScore
}

func GetScore(grid [][]int) int {
	score := 0
	for i := 0; i < len(grid); i++ {
		score += GetRowScore(grid, i)
	}
	return score
}

func matrixScore(grid [][]int) int {
	MaximizeRows(grid)
	MaximizeCols(grid)
	return GetScore(grid)
}
